### PR TITLE
Modified code for patches and patchJson6902 which will throw error if target is not found

### LINKED ISSUE #4379: patches should throw error on missing targets
**Describe the bug**

If `kustomize build` is run with a patchesJson6902 section specifying a nonexistant target it still reports success and outputs its resources as if it worked

**Files that can reproduce the issue**

kustomization.yaml (note the 'v2' in the version field)

```yaml
kind: ""
apiversion: ""
resources:
- manifests/servicemonitor.yaml
patchesJson6902:
- target:
    group: monitoring.coreos.com
    kind: ServiceMonitor
    name: starboard-exporter
    namespace: starboard
    version: v2

  path: jsonpatches/patch.0.yaml
```

manifests/servicemonitor.yaml

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    app: starboard-exporter
  name: starboard-exporter
  namespace: starboard
spec:
  endpoints:
  - path: /metrics
    port: metrics
  selector:
    matchLabels:
      app.kubernetes.io/instance: starboard-exporter
      app.kubernetes.io/name: starboard-exporter
```

jsonpatches/patch.0.yaml

```yaml
- op: add
  path: /metadata/labels/release
  value: kube-prometheus-stack
```

**Expected output**

Some kind of error e.g, "WARNING: patchesJson6902[0] target not found for $target - please check kustomized resources for correctness" or "ERROR: patchesJson6902[0] target not found for $target! Bailing out."
An exit code > 0 to signal that something went wrong
Some suggestions of partial matches e.g., "Found $similartarget, did you mean $targetdiff?"

**Actual output**

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    app: starboard-exporter
  name: starboard-exporter
  namespace: starboard
spec:
  endpoints:
  - path: /metrics
    port: metrics
  selector:
    matchLabels:
      app.kubernetes.io/instance: starboard-exporter
      app.kubernetes.io/name: starboard-exporter
```

**Kustomize version**

```
❯ kustomize version
{Version:kustomize/v4.4.1 GitCommit:b2d65ddc98e09187a8e38adc27c30bab078c1dbf BuildDate:2021-11-11T23:27:14Z GoOs:darwin GoArch:arm64}
```

**Platform**

```
❯ sw_vers
ProductName:    macOS
ProductVersion: 11.6.2
BuildVersion:   20G314
```

### PR DESCRIPTION
**Pull request by laxmikantbpandhare**
Fix for `patchesJson6902`, `patches` is completed. This fix verifies whether is `resources` filed has any data or not. If it is empty then there is definitely a missing target. 

`patchStrategicMerge` is already throwing an error if the `target` is missing.

This will fix #4379

### CODE REVIEW

File: api/internal/builtins/PatchJson6902Transformer.go
Comment (laxmikantbpandhare) [reply api/internal/builtins/PatchJson6902Transformer.go:80]: @yuwenma - I will remove this tab. This was the file added previously which is not needed. Right now, I am adding the test cases. I will commit both together.

Comment (laxmikantbpandhare) [reply api/internal/builtins/PatchJson6902Transformer.go]: Updated and removed the space.

**Comment by yuwenma** [comment]
/lgtm

This is to kick-off the CI test

**Comment by laxmikantbpandhare** [comment]
@KnVerey @natasha41575 - Could you please take a look at this.

**Comment by natasha41575** [comment]
Could you please rebase?

**Comment by laxmikantbpandhare** [comment]
@natasha41575 - I did the rebase.

**Comment by laxmikantbpandhare** [comment]
Also, I squashed commits too.

**Comment by laxmikantbpandhare** [comment]
@KnVerey @natasha41575 - PR is failing with `Lint` errors and it has nothing to do with my changes.

```
Error: api/krusty/openapitests/openapicustomschema_test.go:10:2: SA1019: package io/ioutil is deprecated: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
```

Do you want me to check this as well? Looks like `io/ioutil` got deprecated.

**Comment by KnVerey** [comment]
don't worry about the ioutil warnings; they're fixed in https://github.com/kubernetes-sigs/kustomize/pull/4754

**Comment by laxmikantbpandhare** [comment]
oh ho sorry just submitted issue on `kustomize` :(

**Comment by annasong20** [comment]
#4754 merged, so if you rebase your Lint error should go away.

**Comment by laxmikantbpandhare** [comment]
> #4754 merged, so if you rebase your Lint error should go away.

@annasong20 - I did the rebase, could you please approve the workflow? It needs approval for the execution of jobs.

File: plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go:86]: AH .. let me update it.

Comment (laxmikantbpandhare) [reply plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go:86]: Updated it.

File: plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go]: I removed "Bailing out" completely.

**Comment by laxmikantbpandhare** [comment]
@annasong20 Updated all the requested changes and also squashed the commits.

**Comment by annasong20** [comment]
/lgtm
I don't have merge permissions, so will defer to @natasha41575. 
btw, great use of `types.Selector.String`!

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (natasha41575) [inline plugin/builtin/patchtransformer/PatchTransformer.go:124]:
```suggestion
		return fmt.Errorf("patches target not found for %s", p.Target)
```

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (natasha41575) [inline plugin/builtin/patchtransformer/PatchTransformer.go:106]:
```suggestion
		return fmt.Errorf("patches target not found for %s", p.Target)
```

File: plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go
Comment (natasha41575) [inline plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go:87]:
```suggestion
		return fmt.Errorf("patchesJson6902 target not found for %s", p.Target)
```

File: api/krusty/extendedpatch_test.go
Comment (natasha41575) [inline api/krusty/extendedpatch_test.go:1108]: What is the `:a=:l=t.` at the end? Can we remove it from the error message?

**Review by natasha41575 — COMMENTED** [review]
Some nits, mostly LGTM. 

Sorry for approving and then unapproving this, caught these afterwards

File: api/krusty/extendedpatch_test.go
Comment (laxmikantbpandhare) [reply api/krusty/extendedpatch_test.go:1108]: sure, I will add only Target's name in the error message.

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchtransformer/PatchTransformer.go]: remmoved dot

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchtransformer/PatchTransformer.go]: removed dot

File: api/krusty/extendedpatch_test.go
Comment (laxmikantbpandhare) [reply api/krusty/extendedpatch_test.go]: @natasha41575 - Updated PR with the requested changes. PTAL.

File: plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go]: removed dot

File: api/krusty/extendedpatch_test.go
Comment (natasha41575) [reply api/krusty/extendedpatch_test.go]: > Just a quick FYI: the a stands for annotationSelector and l for labelSelector that can be specified in target

Thanks for the explanation! Still probably not something we want to surface to users in such an incomprehensible way, so I'd prefer the error just include GVKN/N.

Comment (laxmikantbpandhare) [reply api/krusty/extendedpatch_test.go]: @natasha41575 @annasong20 - Updated PR as per the above discussion. PTAL.

File: plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go:87]: `ResId` definitely makes sense.

Comment (laxmikantbpandhare) [reply plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go]: @natasha41575 - Updated the PR and also squashed the commits.

**Review by natasha41575 — APPROVED** [review]
/lgtm

**Comment by laxmikantbpandhare** [comment]
> /hold for discussion

@natasha41575 sure. Could you please do me one favor? This is my first PR that's why need the approval to run the job workflow. Could you please approve once, just to check everything is working.

**Comment by laxmikantbpandhare** [comment]
Great, all jobs ran fine.

**Comment by laxmikantbpandhare** [comment]
@natasha41575 @KnVerey - It's been a while now for this PR. Could you please discuss on this and merge it.

**Comment by lwille** [comment]
Hi, just want to say thank you to @laxmikantbpandhare for pushing this topic this far.
It would be a shame if that effort was lost - what's the status of the discussion, @natasha41575?

**Comment by Arabus** [comment]
Could you rebase the MR please? And also maybe consider https://github.com/kubernetes-sigs/kustomize/issues/4379#issuecomment-1344205011 i.e., make this apply be default but add the possibility to disable the check?

**Comment by laxmikantbpandhare** [comment]
> Could you rebase the MR please? And also maybe consider [#4379 (comment)](https://github.com/kubernetes-sigs/kustomize/issues/4379#issuecomment-1344205011) i.e., make this apply be default but add the possibility to disable the check?

@Arabus - I was on vacation due to which I missed the comments. I rebased this PR.

@KnVerey @natasha41575 - Could you please take a look into it.

**Comment by annasong20** [comment]
@laxmikantbpandhare Sorry for taking so long to get back to you. We just re-discussed #4379 and concluded https://github.com/kubernetes-sigs/kustomize/issues/4379#issuecomment-1480440353. If you'd like to continue the implementation in this PR, I'll try my best to give you timely reviews.

**Comment by laxmikantbpandhare** [comment]
> @laxmikantbpandhare Sorry for taking so long to get back to you. We just re-discussed #4379 and concluded [#4379 (comment)](https://github.com/kubernetes-sigs/kustomize/issues/4379#issuecomment-1480440353). If you'd like to continue the implementation in this PR, I'll try my best to give you timely reviews.

@annasong20 - thank you for your comment. I will start looking into it.

**Comment by laxmikantbpandhare** [comment]
As of now, Added as a log warning. I would like to know more about the **stderr** requirement for patchJsonTransformer. patchjson6902Transformer I kept it as it is as it ll be deprecated soon.

**Comment by laxmikantbpandhare** [comment]
Tes Jobs are failing as test cases not updated yet.

File: plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go
Comment (annasong20) [inline plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go:85]: This is the correct behavior!

**Review by annasong20 — CHANGES_REQUESTED** [review]
Thank you for all of your work on this, @laxmikantbpandhare!

I gave a high-level review of the current code behavior. Lmk if anything is unclear. Once we've fulfilled the behavior agreed upon in #4379 and added tests, I'll do another pass for details.

File: plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go:85]: Thank you for your review. I will keep patchjson6902transformer changes as it is. I will work on other suggestions.

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchtransformer/PatchTransformer.go:105]: Updated as below:

		fmt.Fprintf(os.Stderr, "%v\n", "Warning: patches target not found for Target")

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchtransformer/PatchTransformer.go:104]: totally agree.

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchtransformer/PatchTransformer.go:124]: Updated same as above.

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (laxmikantbpandhare) [reply plugin/builtin/patchtransformer/PatchTransformer.go:104]: Updated as below:

	if p.Options["allowNoTargetMatch"] && len(selected) == 0 {

Comment (laxmikantbpandhare) [reply plugin/builtin/patchtransformer/PatchTransformer.go:104]: @annasong20 - We can create a PatchTransformer struct under `api/types`. So, do you mean only the `options` field can be moved out of it and add it in `api/types`? Please let me know if I am wrong.

Comment (annasong20) [reply plugin/builtin/patchtransformer/PatchTransformer.go:104]: Good question! I only need `options` to be a struct under `api/types`, but if you'd like you can move the PatchTransformer arguments into a `PatchArgs`. The latter is more work, but I think easier to read in the long run.

Comment (annasong20) [reply plugin/builtin/patchtransformer/PatchTransformer.go:104]: I see your comment, but this doesn't seem to be reflected in the code? Can you commit your changes?

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (annasong20) [reply plugin/builtin/patchtransformer/PatchTransformer.go:105]: I don't see this change in the code.

Comment (annasong20) [inline plugin/builtin/patchtransformer/PatchTransformer.go:105]: Can we print for users the "Target" that we're looking for?

As a user with a complex Kustomization, I'd be confused and need to debug which patch target is missing.

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (annasong20) [reply plugin/builtin/patchtransformer/PatchTransformer.go:124]: Don't see the change in code.

File: plugin/builtin/patchtransformer/PatchTransformer.go
Comment (annasong20) [inline plugin/builtin/patchtransformer/PatchTransformer.go:106]: We still need to return an error if `allowNoTargetMatch` is not set?

File: api/krusty/extendedpatch_test.go
Comment (annasong20) [inline api/krusty/extendedpatch_test.go:1085]: Can we use the existing files in `makeCommonFileForExtendedPatchTest` for this test?

File: api/krusty/extendedpatch_test.go
Comment (annasong20) [inline api/krusty/extendedpatch_test.go:1070]: We don't seem to have tests for
* `allowNoTargetMatch`
* json patches under the `patches` field

**Review by annasong20 — CHANGES_REQUESTED** [review]
@laxmikantbpandhare Sorry for taking so long to re-review. I took a closer look this time. Can you address my new comments?

Also, could you rebase?

**Comment by wibed** [comment]
@laxmikantbpandhare 

selfish bump, looking forward to it

**Comment by adoramshoval** [comment]
I continued this PR in #5917

**Pull request closed without merging**
